client/landing/gutenboarding/path.ts

```diff
@@ -189,7 +189,7 @@ export function useAnchorFmParams(): AnchorFmParams {
 	const anchorFmIsNewSite = useAnchorParameter( {
 		queryParamName: 'anchor_is_new_site',
 		locationStateParamName: 'anchorFmIsNewSite',
-		sanitize: sanitizeNumberParam,
+		sanitize: sanitizeAlphaNumeric,
```

Comment:
In my opinion, we shouldn't be carrying around strings like `"true"` and `"false"`. They're too easy to mess up. I would rather convert them to booleans. However, I'm having difficulty doing it while appeasing TypeScript and keeping the flexibility of the `useAnchorParameter()` function.  Using Generics is a great start, but values get "laundered" through the `locationState` object, and I need to narrow down the value I pull out of `locationState[ locationStateParamName ]` to type `<T>`,  which I'm not successful doing.  It has to do with TypeScript not knowing that my `locationStateParamName` will always correspond to pulling out a value of the type that I want, and I'm not sure how to tell it that.  Here is my abandoned diff:

```diff
diff --git a/client/landing/gutenboarding/path.ts b/client/landing/gutenboarding/path.ts
index abda1c3f2c..fae815b8ab 100644
--- a/client/landing/gutenboarding/path.ts
+++ b/client/landing/gutenboarding/path.ts
@@ -54,9 +54,9 @@ export interface GutenLocationStateType {
 	anchorFmPodcastId?: string;
 	anchorFmEpisodeId?: string;
 	anchorFmSpotifyUrl?: string;
-	anchorFmSite?: string;
-	anchorFmPost?: string;
-	anchorFmIsNewSite?: string;
+	anchorFmSite?: number;
+	anchorFmPost?: number;
+	anchorFmIsNewSite?: boolean;
 }
 export type GutenLocationStateKeyType = keyof GutenLocationStateType;
 
@@ -130,13 +130,13 @@ export interface AnchorFmParams {
 	anchorFmPodcastId: string | null;
 	anchorFmEpisodeId: string | null;
 	anchorFmSpotifyUrl: string | null;
-	anchorFmSite: string | null;
-	anchorFmPost: string | null;
-	anchorFmIsNewSite: string | null;
+	anchorFmSite: number | null;
+	anchorFmPost: number | null;
+	anchorFmIsNewSite: boolean | null;
 }
 export function useAnchorFmParams(): AnchorFmParams {
 	const sanitizeAlphaNumeric = ( id: string ) => id.replace( /[^a-zA-Z0-9]/g, '' );
-	const anchorFmPodcastId = useAnchorParameter( {
+	const anchorFmPodcastId = useAnchorParameter< string >( {
 		queryParamName: 'anchor_podcast',
 		locationStateParamName: 'anchorFmPodcastId',
 		sanitize: sanitizeAlphaNumeric,
@@ -146,7 +146,7 @@ export function useAnchorFmParams(): AnchorFmParams {
 	// Reserved characters: !*'();:@&=+$,/?#[]
 	// Unreserved: A-Za-z0-9_.~-    (possibly % as a part of percent-encoding)
 	const sanitizeEpisode = ( id: string ) => id.replace( /[^A-Za-z0-9_.\-~%]/g, '' );
-	const anchorFmEpisodeId = useAnchorParameter( {
+	const anchorFmEpisodeId = useAnchorParameter< string >( {
 		queryParamName: 'anchor_episode',
 		locationStateParamName: 'anchorFmEpisodeId',
 		sanitize: sanitizeEpisode,
@@ -157,7 +157,7 @@ export function useAnchorFmParams(): AnchorFmParams {
 	// Unreserved: A-Za-z0-9_.~-    (possibly % as a part of percent-encoding)
 	const sanitizeShowUrl = ( id: string ) =>
 		id.replace( /[^A-Za-z0-9_.\-~%!*'();:@&=+$,/?#[\]]/g, '' );
-	const anchorFmSpotifyUrl = useAnchorParameter( {
+	const anchorFmSpotifyUrl = useAnchorParameter< string >( {
 		queryParamName: 'spotify_url',
 		locationStateParamName: 'anchorFmSpotifyUrl',
 		sanitize: sanitizeShowUrl,
@@ -167,13 +167,13 @@ export function useAnchorFmParams(): AnchorFmParams {
 	// http://wordpress.com/new?site=181129564&post=5&anchor_podcast=22b6608
 	// We store them as strings for consistency with the other param types
 	// and simplicity in code and type signatures.
-	const sanitizeNumberParam = ( id: string ) => id.replace( /^\D+$/g, '' );
-	const anchorFmSite = useAnchorParameter( {
+	const sanitizeNumberParam = ( id: string ) => parseInt( id.replace( /^\D+$/g, '' ), 10 ) ?? null;
+	const anchorFmSite = useAnchorParameter< number >( {
 		queryParamName: 'site',
 		locationStateParamName: 'anchorFmSite',
 		sanitize: sanitizeNumberParam,
 	} );
-	const anchorFmPost = useAnchorParameter( {
+	const anchorFmPost = useAnchorParameter< number >( {
 		queryParamName: 'post',
 		locationStateParamName: 'anchorFmPost',
 		sanitize: sanitizeNumberParam,
@@ -186,10 +186,12 @@ export function useAnchorFmParams(): AnchorFmParams {
 	// no matching site and redirect us to "/new?anchor_podcast=abcdef0&anchor_episode=1234-123456&anchor_is_new_site=true",
 	// because it found the last episode and wanted to pass that information to us.
 	// In this case, we don't need to ask the backend again after restarting gutenboarding.
-	const anchorFmIsNewSite = useAnchorParameter( {
+	const stringToBool = ( id: string ) =>
+		'true'.localeCompare( id, undefined, { sensitivity: 'accent' } ) === 0; // Case-insensitive compare to string "true"
+	const anchorFmIsNewSite = useAnchorParameter< boolean >( {
 		queryParamName: 'anchor_is_new_site',
 		locationStateParamName: 'anchorFmIsNewSite',
-		sanitize: sanitizeAlphaNumeric,
+		sanitize: stringToBool,
 	} );
 
 	return {
@@ -217,16 +219,16 @@ export function useAnchorFmParams(): AnchorFmParams {
   })
   Looks for the value first in location state, then if it can't be found, looks in the query parameter.
 */
-interface UseAnchorParameterType {
+interface UseAnchorParameterType< T > {
 	queryParamName: string;
 	locationStateParamName: GutenLocationStateKeyType;
-	sanitize: ( arg0: string ) => string;
+	sanitize: ( arg0: string ) => T;
 }
-function useAnchorParameter( {
+function useAnchorParameter< T >( {
 	queryParamName,
 	locationStateParamName,
 	sanitize,
-}: UseAnchorParameterType ): string | null {
+}: UseAnchorParameterType< T > ): T | null {
 	const { state: locationState = {}, search } = useLocation< GutenLocationStateType >();
 
 	// Feature flag 'anchor-fm-dev' is required for anchor podcast id to be read
```

I have made the requested changes, but it's not true that `true` is our one valid value, since `anchorFmIsNewSite` is checked in a boolean context here:
https://github.com/Automattic/wp-calypso/blob/c281b005cd04daaa2553c837678e7d3640511da2/client/landing/gutenboarding/hooks/use-detect-matching-anchor-site.ts#L38

I'm guessing you would like this change to be made as well?

```diff
diff --git a/client/landing/gutenboarding/hooks/use-detect-matching-anchor-site.ts b/client/landing/gutenboarding/hooks/use-detect-matching-anchor-site.ts
index cabf1e8638..567d491e05 100644
--- a/client/landing/gutenboarding/hooks/use-detect-matching-anchor-site.ts
+++ b/client/landing/gutenboarding/hooks/use-detect-matching-anchor-site.ts
@@ -35,7 +35,7 @@ export default function useDetectMatchingAnchorSite(): boolean {
 
        React.useEffect( () => {
                // Must be a logged-in user on anchor FM to check
-               if ( ! isAnchorFm || ! currentUserExists || anchorFmIsNewSite ) {
+               if ( ! isAnchorFm || ! currentUserExists || "true" === anchorFmIsNewSite ) {
                        setIsLoading( false );
                        return;
                }
```

Reply:
Jeez, yeah – let's go with what we've got. You're right, I thought the strings were weird too, just... less weird? And not worth the rabbit hole above, no.